File: packages/daemon/src/envelope.js
// @ts-check

/**
 * Minimal CBOR codec for the engo envelope protocol.
 *
 * Envelopes are 4-element CBOR arrays: [handle, verb, payload, nonce].
 * Frames are CBOR byte strings wrapping encoded envelopes.
 *
 * This codec is intentionally minimal — it only handles the types used
 * by the envelope protocol (unsigned/negative integers, byte strings,
 * text strings, arrays, maps) and does not attempt to be a general-purpose
 * CBOR library.
 */

// CBOR major types
const CBOR_UINT = 0;
const CBOR_NEGINT = 1;
const CBOR_BYTES = 2;
const CBOR_TEXT = 3;
const CBOR_ARRAY = 4;
// const CBOR_MAP = 5;

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

// ---------------------------------------------------------------------------
// CBOR encoding
// ---------------------------------------------------------------------------

/**
 * Append a CBOR head (major type + argument) to a byte list.
 * @param {number[]} buf - mutable byte array
 * @param {number} major - major type (0-7)
 * @param {number} n - argument value
 */
const cborAppendHead = (buf, major, n) => {
  const m = major << 5;
  if (n < 24) {
    buf.push(m | n);
  } else if (n <= 0xff) {
    buf.push(m | 24, n);
  } else if (n <= 0xffff) {
    buf.push(m | 25, (n >> 8) & 0xff, n & 0xff);
  } else if (n <= 0xffffffff) {
    buf.push(
      m | 26,
      (n >> 24) & 0xff,
      (n >> 16) & 0xff,
      (n >> 8) & 0xff,
      n & 0xff,
    );
  } else {
    // For values > 32 bits, we'd need BigInt. Handle sizes are small.
    throw new Error(`CBOR value too large: ${n}`);
  }
};

/**
 * Encode an integer (signed).
 * @param {number[]} buf
 * @param {number} n
 */
const cborAppendInt = (buf, n) => {
  if (n >= 0) {
    cborAppendHead(buf, CBOR_UINT, n);
  } else {
    cborAppendHead(buf, CBOR_NEGINT, -1 - n);
  }
};

/**
 * Encode a byte string.
 * @param {number[]} buf
 * @param {Uint8Array} data
 */
const cborAppendBytes = (buf, data) => {
  cborAppendHead(buf, CBOR_BYTES, data.length);
  for (let i = 0; i < data.length; i += 1) {
    buf.push(data[i]);
  }
};

/**
 * Encode a text string.
 * @param {number[]} buf
 * @param {string} s
 */
const cborAppendText = (buf, s) => {
  const encoded = textEncoder.encode(s);
  cborAppendHead(buf, CBOR_TEXT, encoded.length);
  for (let i = 0; i < encoded.length; i += 1) {
    buf.push(encoded[i]);
  }
};

/**
 * @typedef {object} Envelope
 * @property {number} handle
 * @property {string} verb
 * @property {Uint8Array} payload
 * @property {number} nonce
 */

/**
 * Encode an envelope as a CBOR 4-element array.
 * @param {Envelope} env
 * @returns {Uint8Array}
 */
export const encodeEnvelope = env => {
  /** @type {number[]} */
  const buf = [];
  cborAppendHead(buf, CBOR_ARRAY, 4);
  cborAppendInt(buf, env.handle);
  cborAppendText(buf, env.verb);
  cborAppendBytes(buf, env.payload || new Uint8Array(0));
  cborAppendInt(buf, env.nonce || 0);
  return new Uint8Array(buf);
};
harden(encodeEnvelope);

/**
 * Encode a CBOR frame: a byte string wrapping the given data.
 * @param {Uint8Array} data
 * @returns {Uint8Array}
 */
export const encodeFrame = data => {
  /** @type {number[]} */
  const buf = [];
  cborAppendBytes(buf, data);
  return new Uint8Array(buf);
};
harden(encodeFrame);

// ---------------------------------------------------------------------------
// CBOR decoding
// ---------------------------------------------------------------------------

/**
 * A simple cursor for reading CBOR from a Uint8Array.
 * @param {Uint8Array} data
 * @returns {{ pos: number, data: Uint8Array }}
 */
const makeCursor = data => ({ pos: 0, data });

/**
 * Read a CBOR head from the cursor.
 * @param {{ pos: number, data: Uint8Array }} cursor
 * @returns {{ major: number, value: number }}
 */
const cborReadHead = cursor => {
  if (cursor.pos >= cursor.data.length) {
    throw new Error('CBOR: unexpected end of input');
  }
  const initial = cursor.data[cursor.pos];
  cursor.pos += 1;
  const major = initial >> 5;
  const info = initial & 0x1f;
  if (info < 24) {
    return { major, value: info };
  }
  let size;
  if (info === 24) size = 1;
  else if (info === 25) size = 2;
  else if (info === 26) size = 4;
  else if (info === 27) size = 8;
  else throw new Error(`CBOR: unsupported additional info ${info}`);
  let value = 0;
  for (let i = 0; i < size; i += 1) {
    value = value * 256 + cursor.data[cursor.pos + i];
  }
  cursor.pos += size;
  return { major, value };
};

/**
 * Read a CBOR integer (signed).
 * @param {{ pos: number, data: Uint8Array }} cursor
 * @returns {number}
 */
const cborReadInt = cursor => {
  const { major, value } = cborReadHead(cursor);
  if (major === CBOR_UINT) return value;
  if (major === CBOR_NEGINT) return -1 - value;
  throw new Error(`CBOR: expected int, got major ${major}`);
};

/**
 * Read a CBOR byte string.
 * @param {{ pos: number, data: Uint8Array }} cursor
 * @returns {Uint8Array}
 */
const cborReadBytes = cursor => {
  const { major, value } = cborReadHead(cursor);
  if (major !== CBOR_BYTES) {
    throw new Error(`CBOR: expected bytes (major 2), got major ${major}`);
  }
  const result = cursor.data.subarray(cursor.pos, cursor.pos + value);
  cursor.pos += value;
  return new Uint8Array(result);
};

/**
 * Read a CBOR text string.
 * @param {{ pos: number, data: Uint8Array }} cursor
 * @returns {string}
 */
const cborReadText = cursor => {
  const { major, value } = cborReadHead(cursor);
  if (major !== CBOR_TEXT) {
    throw new Error(`CBOR: expected text (major 3), got major ${major}`);
  }
  const result = textDecoder.decode(
    cursor.data.subarray(cursor.pos, cursor.pos + value),
  );
  cursor.pos += value;
  return result;
};

/**
 * Read a CBOR array header.
 * @param {{ pos: number, data: Uint8Array }} cursor
 * @returns {number} - number of elements
 */
const cborReadArrayHeader = cursor => {
  const { major, value } = cborReadHead(cursor);
  if (major !== CBOR_ARRAY) {
    throw new Error(`CBOR: expected array (major 4), got major ${major}`);
  }
  return value;
};

/**
 * Decode a CBOR frame (byte string) from raw bytes.
 * @param {Uint8Array} frameData
 * @returns {Uint8Array} - the inner content
 */
export const decodeFrame = frameData => {
  const cursor = makeCursor(frameData);
  return cborReadBytes(cursor);
};
harden(decodeFrame);

/**
 * Decode an envelope from a CBOR 4-element array.
 * @param {Uint8Array} data
 * @returns {Envelope}
 */
export const decodeEnvelope = data => {
  const cursor = makeCursor(data);
  const n = cborReadArrayHeader(cursor);
  if (n !== 3 && n !== 4) {
    throw new Error(`Envelope: expected 3 or 4 elements, got ${n}`);
  }
  const handle = cborReadInt(cursor);
  const verb = cborReadText(cursor);
  const payload = cborReadBytes(cursor);
  const nonce = n === 4 ? cborReadInt(cursor) : 0;
  return harden({ handle, verb, payload, nonce });
};
harden(decodeEnvelope);

// ---------------------------------------------------------------------------
// Streaming: read CBOR frames from a Node.js readable stream
// ---------------------------------------------------------------------------

/**
 * Read exactly `n` bytes from a Node.js readable stream.
 * @param {import('stream').Readable} stream
 * @param {number} n
 * @returns {Promise<Uint8Array | null>}
 */
const readExactly = (stream, n) => {
  return new Promise((resolve, reject) => {
    const chunks = /** @type {Buffer[]} */ ([]);
    let remaining = n;

    const onReadable = () => {
      while (remaining > 0) {
        const chunk = stream.read(Math.min(remaining, stream.readableLength));
        if (chunk === null) {
          return; // Wait for more data
        }
        chunks.push(chunk);
        remaining -= chunk.length;
      }
      cleanup();
      resolve(new Uint8Array(Buffer.concat(chunks)));
    };

    const onEnd = () => {
      cleanup();
      if (remaining === n) {
        resolve(null); // Clean EOF
      } else {
        reject(new Error('CBOR: unexpected EOF in frame'));
      }
    };

    const onError = (/** @type {Error} */ err) => {
      cleanup();
      reject(err);
    };

    const cleanup = () => {
      stream.removeListener('readable', onReadable);
      stream.removeListener('end', onEnd);
      stream.removeListener('error', onError);
    };

    stream.on('readable', onReadable);
    stream.on('end', onEnd);
    stream.on('error', onError);

    // Try reading immediately in case data is already buffered.
    onReadable();
  });
};

/**
 * Read one CBOR byte-string frame from a Node.js readable stream.
 * Returns the inner content bytes, or null on EOF.
 * @param {import('stream').Readable} stream
 * @returns {Promise<Uint8Array | null>}
 */
export const readFrameFromStream = async stream => {
  // Read the CBOR byte-string header.
  const firstByte = await readExactly(stream, 1);
  if (firstByte === null) return null;

  const major = firstByte[0] >> 5;
  if (major !== CBOR_BYTES) {
    throw new Error(
      `CBOR frame: expected byte string (major 2), got major ${major}`,
    );
  }
  const info = firstByte[0] & 0x1f;

  let length;
  if (info < 24) {
    length = info;
  } else if (info === 24) {
    const b = await readExactly(stream, 1);
    if (b === null) throw new Error('CBOR: unexpected EOF in frame header');
    length = b[0];
  } else if (info === 25) {
    const b = await readExactly(stream, 2);
    if (b === null) throw new Error('CBOR: unexpected EOF in frame header');
    length = (b[0] << 8) | b[1];
  } else if (info === 26) {
    const b = await readExactly(stream, 4);
    if (b === null) throw new Error('CBOR: unexpected EOF in frame header');
    length = (b[0] << 24) | (b[1] << 16) | (b[2] << 8) | b[3];
  } else {
    throw new Error(`CBOR frame: unsupported length info ${info}`);
  }

  if (length === 0) return new Uint8Array(0);

  const content = await readExactly(stream, length);
  if (content === null) {
    throw new Error('CBOR: unexpected EOF in frame content');
  }
  return content;
};
harden(readFrameFromStream);

/**
 * Write a CBOR byte-string frame to a Node.js writable stream.
 * @param {import('stream').Writable} stream
 * @param {Uint8Array} data
 * @returns {Promise<void>}
 */
export const writeFrameToStream = (stream, data) => {
  const frame = encodeFrame(data);
  return new Promise((resolve, reject) => {
    stream.write(frame, err => {
      if (err) reject(err);
      else resolve(undefined);
    });
  });
};
harden(writeFrameToStream);

File: rust/endo/xsnap/src/envelope.rs
//! CBOR envelope protocol for worker ↔ supervisor communication.
//!
//! Envelopes are 4-element CBOR arrays: [handle, verb, payload, nonce].
//! Frames are CBOR byte-string wrappers around serialized envelopes.
//!
//! This is a self-contained copy of the codec from `rust/endo/src/codec.rs`,
//! without tokio async dependencies. The xsnap worker runs a synchronous
//! event loop (XS is single-threaded), so async I/O is not needed.

use std::io::{self, Read, Write};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// Worker/daemon handle identifier.
pub type Handle = i64;

/// An envelope on the wire.
#[derive(Debug, Clone)]
pub struct Envelope {
    pub handle: Handle,
    pub verb: String,
    pub payload: Vec<u8>,
    pub nonce: i64,
}

// ---------------------------------------------------------------------------
// CBOR encoding helpers
// ---------------------------------------------------------------------------

const CBOR_UINT: u8 = 0;
const CBOR_NEGINT: u8 = 1;
const CBOR_BYTES: u8 = 2;
const CBOR_TEXT: u8 = 3;
const CBOR_ARRAY: u8 = 4;

fn cbor_append_head(buf: &mut Vec<u8>, major: u8, n: u64) {
    let m = major << 5;
    if n < 24 {
        buf.push(m | n as u8);
    } else if n <= 0xff {
        buf.push(m | 24);
        buf.push(n as u8);
    } else if n <= 0xffff {
        buf.push(m | 25);
        buf.extend_from_slice(&(n as u16).to_be_bytes());
    } else if n <= 0xffff_ffff {
        buf.push(m | 26);
        buf.extend_from_slice(&(n as u32).to_be_bytes());
    } else {
        buf.push(m | 27);
        buf.extend_from_slice(&n.to_be_bytes());
    }
}

fn cbor_append_int(buf: &mut Vec<u8>, n: i64) {
    if n >= 0 {
        cbor_append_head(buf, CBOR_UINT, n as u64);
    } else {
        cbor_append_head(buf, CBOR_NEGINT, (-1 - n) as u64);
    }
}

fn cbor_append_bytes(buf: &mut Vec<u8>, data: &[u8]) {
    cbor_append_head(buf, CBOR_BYTES, data.len() as u64);
    buf.extend_from_slice(data);
}

fn cbor_append_text(buf: &mut Vec<u8>, s: &str) {
    cbor_append_head(buf, CBOR_TEXT, s.len() as u64);
    buf.extend_from_slice(s.as_bytes());
}

// ---------------------------------------------------------------------------
// CBOR decoding helpers
// ---------------------------------------------------------------------------

struct Cursor<'a> {
    data: &'a [u8],
    pos: usize,
}

impl<'a> Cursor<'a> {
    fn new(data: &'a [u8]) -> Self {
        Cursor { data, pos: 0 }
    }

    fn read_head(&mut self) -> io::Result<(u8, u64)> {
        if self.pos >= self.data.len() {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                "CBOR: unexpected end of input",
            ));
        }
        let initial = self.data[self.pos];
        self.pos += 1;
        let major = initial >> 5;
        let info = initial & 0x1f;
        if info < 24 {
            return Ok((major, info as u64));
        }
        let size = match info {
            24 => 1,
            25 => 2,
            26 => 4,
            27 => 8,
            _ => {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!("CBOR: unsupported info {info}"),
                ))
            }
        };
        if self.pos + size > self.data.len() {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                "CBOR: truncated head",
            ));
        }
        let mut value: u64 = 0;
        for i in 0..size {
            value = (value << 8) | self.data[self.pos + i] as u64;
        }
        self.pos += size;
        Ok((major, value))
    }

    fn read_int(&mut self) -> io::Result<i64> {
        let (major, value) = self.read_head()?;
        match major {
            0 => Ok(value as i64),
            1 => Ok(-(value as i64) - 1),
            _ => Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("CBOR: expected int, got major {major}"),
            )),
        }
    }

    fn read_bytes(&mut self) -> io::Result<Vec<u8>> {
        let (major, len) = self.read_head()?;
        if major != CBOR_BYTES {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("CBOR: expected bytes, got major {major}"),
            ));
        }
        let len = len as usize;
        if self.pos + len > self.data.len() {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                "CBOR: truncated bytes",
            ));
        }
        let result = self.data[self.pos..self.pos + len].to_vec();
        self.pos += len;
        Ok(result)
    }

    fn read_text(&mut self) -> io::Result<String> {
        let (major, len) = self.read_head()?;
        if major != CBOR_TEXT {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("CBOR: expected text, got major {major}"),
            ));
        }
        let len = len as usize;
        if self.pos + len > self.data.len() {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                "CBOR: truncated text",
            ));
        }
        let s = std::str::from_utf8(&self.data[self.pos..self.pos + len])
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
        self.pos += len;
        Ok(s.to_string())
    }

    fn read_array_header(&mut self) -> io::Result<u64> {
        let (major, len) = self.read_head()?;
        if major != CBOR_ARRAY {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("CBOR: expected array, got major {major}"),
            ));
        }
        Ok(len)
    }
}

// ---------------------------------------------------------------------------
// Frame I/O
// ---------------------------------------------------------------------------

const MAX_FRAME_SIZE: usize = 16 * 1024 * 1024; // 16 MiB

/// Read one CBOR byte-string frame. Returns None on EOF.
pub fn read_frame(r: &mut impl Read) -> io::Result<Option<Vec<u8>>> {
    let mut first = [0u8; 1];
    match r.read_exact(&mut first) {
        Ok(()) => {}
        Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => return Ok(None),
        Err(e) => return Err(e),
    }
    let major = first[0] >> 5;
    if major != CBOR_BYTES {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("frame: expected bytes major 2, got {major}"),
        ));
    }
    let info = first[0] & 0x1f;
    let len: usize = if info < 24 {
        info as usize
    } else if info == 24 {
        let mut b = [0u8; 1];
        r.read_exact(&mut b)?;
        b[0] as usize
    } else if info == 25 {
        let mut b = [0u8; 2];
        r.read_exact(&mut b)?;
        ((b[0] as usize) << 8) | b[1] as usize
    } else if info == 26 {
        let mut b = [0u8; 4];
        r.read_exact(&mut b)?;
        ((b[0] as usize) << 24)
            | ((b[1] as usize) << 16)
            | ((b[2] as usize) << 8)
            | b[3] as usize
    } else {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("frame: unsupported length info {info}"),
        ));
    };
    if len > MAX_FRAME_SIZE {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("frame: size {len} exceeds maximum {MAX_FRAME_SIZE}"),
        ));
    }
    let mut buf = vec![0u8; len];
    if len > 0 {
        r.read_exact(&mut buf)?;
    }
    Ok(Some(buf))
}

/// Write a CBOR byte-string frame.
pub fn write_frame(w: &mut impl Write, data: &[u8]) -> io::Result<()> {
    let mut header = Vec::new();
    cbor_append_head(&mut header, CBOR_BYTES, data.len() as u64);
    w.write_all(&header)?;
    w.write_all(data)?;
    w.flush()
}

// ---------------------------------------------------------------------------
// Envelope encoding/decoding
// ---------------------------------------------------------------------------

/// Encode an envelope as a CBOR 4-element array.
pub fn encode_envelope(env: &Envelope) -> Vec<u8> {
    let mut buf = Vec::new();
    cbor_append_head(&mut buf, CBOR_ARRAY, 4);
    cbor_append_int(&mut buf, env.handle);
    cbor_append_text(&mut buf, &env.verb);
    cbor_append_bytes(&mut buf, &env.payload);
    cbor_append_int(&mut buf, env.nonce);
    buf
}

/// Decode an envelope from a CBOR 3- or 4-element array.
pub fn decode_envelope(data: &[u8]) -> io::Result<Envelope> {
    let mut c = Cursor::new(data);
    let n = c.read_array_header()?;
    if n != 3 && n != 4 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("envelope: expected 3 or 4 elements, got {n}"),
        ));
    }
    let handle = c.read_int()?;
    let verb = c.read_text()?;
    let payload = c.read_bytes()?;
    let nonce = if n == 4 { c.read_int()? } else { 0 };
    Ok(Envelope {
        handle,
        verb,
        payload,
        nonce,
    })
}

// ---------------------------------------------------------------------------
// Convenience: read/write envelope as a framed message
// ---------------------------------------------------------------------------

/// Read one envelope from a framed stream. Returns None on EOF.
pub fn read_envelope(r: &mut impl Read) -> io::Result<Option<Envelope>> {
    match read_frame(r)? {
        Some(data) => Ok(Some(decode_envelope(&data)?)),
        None => Ok(None),
    }
}

/// Write one envelope as a framed message.
pub fn write_envelope(w: &mut impl Write, env: &Envelope) -> io::Result<()> {
    let data = encode_envelope(env);
    write_frame(w, &data)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip_envelope() {
        let env = Envelope {
            handle: 42,
            verb: "deliver".to_string(),
            payload: b"hello".to_vec(),
            nonce: 7,
        };
        let encoded = encode_envelope(&env);
        let decoded = decode_envelope(&encoded).unwrap();
        assert_eq!(decoded.handle, 42);
        assert_eq!(decoded.verb, "deliver");
        assert_eq!(decoded.payload, b"hello");
        assert_eq!(decoded.nonce, 7);
    }

    #[test]
    fn round_trip_frame() {
        let data = b"test payload";
        let mut buf = Vec::new();
        write_frame(&mut buf, data).unwrap();

        let mut cursor = io::Cursor::new(&buf);
        let read_back = read_frame(&mut cursor).unwrap().unwrap();
        assert_eq!(read_back, data);
    }

    #[test]
    fn round_trip_framed_envelope() {
        let env = Envelope {
            handle: 1,
            verb: "init".to_string(),
            payload: vec![],
            nonce: 0,
        };
        let mut buf = Vec::new();
        write_envelope(&mut buf, &env).unwrap();

        let mut cursor = io::Cursor::new(&buf);
        let decoded = read_envelope(&mut cursor).unwrap().unwrap();
        assert_eq!(decoded.handle, 1);
        assert_eq!(decoded.verb, "init");
        assert!(decoded.payload.is_empty());
    }

    #[test]
    fn negative_handle() {
        let env = Envelope {
            handle: -1,
            verb: "error".to_string(),
            payload: b"oops".to_vec(),
            nonce: -5,
        };
        let encoded = encode_envelope(&env);
        let decoded = decode_envelope(&encoded).unwrap();
        assert_eq!(decoded.handle, -1);
        assert_eq!(decoded.nonce, -5);
    }

    #[test]
    fn eof_returns_none() {
        let buf: Vec<u8> = vec![];
        let mut cursor = io::Cursor::new(&buf);
        assert!(read_frame(&mut cursor).unwrap().is_none());
        assert!(read_envelope(&mut cursor).unwrap().is_none());
    }

    #[test]
    fn three_element_envelope() {
        // Build a 3-element envelope manually (no nonce)
        let mut buf = Vec::new();
        cbor_append_head(&mut buf, CBOR_ARRAY, 3);
        cbor_append_int(&mut buf, 10);
        cbor_append_text(&mut buf, "deliver");
        cbor_append_bytes(&mut buf, b"data");

        let decoded = decode_envelope(&buf).unwrap();
        assert_eq!(decoded.handle, 10);
        assert_eq!(decoded.verb, "deliver");
        assert_eq!(decoded.payload, b"data");
        assert_eq!(decoded.nonce, 0); // default
    }
}
